californium-core/src/main/java/org/eclipse/californium/core/network/Matcher.java

```diff
@@ -173,6 +177,23 @@ public void sendRequest(Exchange exchange, Request request) {
 		
 		exchangesByMID.put(idByMID, exchange);
 		exchangesByToken.put(idByToken, exchange);
+		
+		if(request.isMulticast()) {
+			executor.schedule(new Runnable() {
+				@Override
+				public void run() {
+						if(exchange.getResponse() == null)
+						{
+							exchange.setComplete();
+						}
+						else
+						{
+							exchange.getCurrentRequest().setCanceled(true);
+							exchange.setComplete();
+						}
+					}
+				}, multicastMaxLifetime, TimeUnit.MILLISECONDS);
+		}
```

Comment:
> More advanced:
> check the exchange, if it has already at least one response. If so, just complete the exchange. If not cancel the request and complete the exchange. I think, this would be the best approach to handle the multiple responses.

I think, the if branches are exchanged according to the recommendation. 
`exchange.getResponse() == null` is the case with no response, where the request was intended to be canceled before complete the exchange.

```
if(exchange.getResponse() == null)
{
    exchange.getCurrentRequest().setCanceled(true);
}
exchange.setComplete();

```

But if it runs "for many days on many different devices", we can also use your solution.
So, do you want to adapt, or should I merge your current solution?

Reply:
I didn't really test the case of no responses to the multicast request. So I made some tests: 
- It appears that I received a timeout notification from the RetransmissionTask of the ReliabilityLayer.
- But after the multicastMaxLifetime I still have some current request which are not canceled.

So I think the best is to implement your last recommendation.